File: src/render/htmlSanitize.ts
import DOMPurify from "dompurify";

// The raw-HTML render policy — the single allowlist that both render surfaces
// apply to raw inline/block HTML embedded in a Markdown/Org block.
//
// WHY a policy and not a parse: lsdoc emits `inline_html`/`raw_html` nodes with
// the source bytes VERBATIM (byte-parity with mldoc — mldoc doesn't sanitize
// either). Sanitizing is a *render-layer* safety decision, so it lives at each
// render boundary, NOT in the parser. Tine has two such boundaries in two
// languages, so this list is MIRRORED in crates/tine-core/src/html_sanitize.rs
// (ammonia, for the static-HTML export). fixtures/html-sanitize-cases.json
// contract-tests that the two agree. Keep them in lockstep — if you add a tag
// or attribute here, add it there and add a fixture.
//
// Threat model: notes aren't self-authored (Syncthing sync, import, paste,
// shared graphs), and in Tauri an injected `onerror=`/`<script>` can call Tine's
// IPC to read/write the whole graph or exfiltrate via `fetch`; the export
// re-publishes raw HTML as served content. So: allowlist tags + attrs, drop all
// event handlers and `style`, and lean on DOMPurify's URI safety (which blocks
// `javascript:` etc.).

/** Tags that survive sanitization — inline text formatting plus a small set of
 *  containers, `<a>`, and `<img>`. Deliberately excludes `<iframe>`, `<script>`,
 *  forms, and anything executable. (The app renders a sandboxed-https `<iframe>`
 *  via a SEPARATE path in `renderRawHtml`, layered above this allowlist.) */
export const RAW_HTML_TAGS = [
  "b", "strong", "i", "em", "u", "ins", "del", "s", "strike", "sub", "sup",
  "mark", "kbd", "abbr", "small", "code", "cite", "q", "span", "br",
  "p", "div", "blockquote", "details", "summary", "a", "img",
];

/** Attributes that survive, across all allowed tags. Note the absence of
 *  `style` (positioning/tracking) and any `on*` handler. */
export const RAW_HTML_ATTRS = [
  "class", "title", "href", "src", "alt", "width", "height", "open",
];

// --- Local-file `<img>` support (opt-in; see localFileSettings + ADR 0019) ---
// The sanitizer strips a `file:`/absolute-path `src`, so a raw-HTML `<img>` pointing
// at a local file loses its src. When the user has opted in, the app matches the
// sanitized `<img>` elements (in document order) back to these scanned paths and
// swaps in a blob URL read over the gated IPC. Pure/string-only so it's unit-tested.

const IMG_RE = /<img\b[^>]*>/gi;
const SRC_RE = /\bsrc\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s>]+))/i;

/** True if `src` is a local filesystem path (not a web / data / blob URL). */
function isLocalSrc(src: string): boolean {
  return /^(?:file:\/\/|\/(?!\/)|[a-zA-Z]:[\\/]|\\\\)/.test(src);
}

/** Normalize a local `<img>` `src` (possibly `file://…`, percent-encoded) to a
 *  filesystem path, or null if it isn't a local path. */
export function localImagePath(src: string): string | null {
  if (!isLocalSrc(src)) return null;
  let p = src.replace(/^file:\/\//i, "");
  try {
    p = decodeURIComponent(p);
  } catch {
    /* leave as-is if it isn't valid percent-encoding */
  }
  // file:///C:/x → /C:/x → C:/x  (drop the leading slash before a drive letter)
  if (/^\/[a-zA-Z]:[\\/]/.test(p)) p = p.slice(1);
  return p;
}

/** For each `<img>` in `text` (document order), its local filesystem path, or null
 *  for a web/data/blob/relative src. Aligns 1:1 with the `<img>` elements the
 *  sanitized HTML yields — `img` is allowlisted, so every one survives sanitizing. */
export function rawHtmlLocalImages(text: string): (string | null)[] {
  const out: (string | null)[] = [];
  for (const m of text.matchAll(IMG_RE)) {
    const s = SRC_RE.exec(m[0]);
    const src = s ? (s[1] ?? s[2] ?? s[3] ?? "") : "";
    out.push(src ? localImagePath(src) : null);
  }
  return out;
}

/** Sanitize a raw-HTML fragment down to {@link RAW_HTML_TAGS}/{@link RAW_HTML_ATTRS}.
 *  Returns a safe HTML string for `innerHTML`. Everything outside the allowlist
 *  (tags, attributes, `javascript:`/`data:text` URIs) is stripped; the text
 *  content of stripped elements is preserved where DOMPurify preserves it. */
export function sanitizeRawHtml(html: string): string {
  return DOMPurify.sanitize(html, {
    ALLOWED_TAGS: RAW_HTML_TAGS,
    ALLOWED_ATTR: RAW_HTML_ATTRS,
    ALLOW_DATA_ATTR: false,
  });
}

File: crates/tine-core/src/html_sanitize.rs
//! Raw-HTML sanitizer for the static-HTML export.
//!
//! This is the Rust half of a MIRRORED policy — the TypeScript half is
//! `src/render/htmlSanitize.ts` (DOMPurify, for the live app). lsdoc emits
//! `raw_html`/`inline_html` nodes with the source bytes verbatim (byte-parity
//! with mldoc); sanitizing is a render-layer safety decision applied at each
//! render boundary, not in the parser. The export re-publishes user HTML as
//! served content, so it needs the same allowlist the app enforces.
//!
//! `fixtures/html-sanitize-cases.json` contract-tests that this and the TS side
//! agree. Keep the two allowlists in lockstep.

use ammonia::Builder;
use std::collections::{HashMap, HashSet};

/// Mirror of `RAW_HTML_TAGS` in `src/render/htmlSanitize.ts`.
const TAGS: &[&str] = &[
    "b",
    "strong",
    "i",
    "em",
    "u",
    "ins",
    "del",
    "s",
    "strike",
    "sub",
    "sup",
    "mark",
    "kbd",
    "abbr",
    "small",
    "code",
    "cite",
    "q",
    "span",
    "br",
    "p",
    "div",
    "blockquote",
    "details",
    "summary",
    "a",
    "img",
];

/// Sanitize a raw-HTML fragment to the shared allowlist. Event handlers,
/// `style`, `<iframe>`/`<script>`, and `javascript:` URIs are stripped;
/// `href`/`src` are limited to safe schemes.
pub fn sanitize(html: &str) -> String {
    let tags: HashSet<&str> = TAGS.iter().copied().collect();

    // `class`/`title` on any tag; `href`/`src`/etc. only where they belong.
    let generic: HashSet<&str> = ["class", "title"].into_iter().collect();
    let mut tag_attrs: HashMap<&str, HashSet<&str>> = HashMap::new();
    tag_attrs.insert("a", ["href"].into_iter().collect());
    tag_attrs.insert(
        "img",
        ["src", "alt", "width", "height"].into_iter().collect(),
    );
    tag_attrs.insert("details", ["open"].into_iter().collect());

    let schemes: HashSet<&str> = ["https", "http", "mailto", "tel", "data"]
        .into_iter()
        .collect();

    Builder::default()
        .tags(tags)
        .generic_attributes(generic)
        .tag_attributes(tag_attrs)
        .url_schemes(schemes)
        .link_rel(Some("noopener noreferrer"))
        .clean(html)
        .to_string()
}

#[cfg(test)]
mod tests {
    use super::sanitize;
    use serde::Deserialize;

    #[derive(Deserialize)]
    struct Case {
        name: String,
        input: String,
        #[serde(rename = "mustContain")]
        must_contain: Vec<String>,
        #[serde(rename = "mustNotContain")]
        must_not_contain: Vec<String>,
    }
    #[derive(Deserialize)]
    struct Cases {
        cases: Vec<Case>,
    }

    // Same fixtures the TS/DOMPurify side runs — this asserts the two render
    // surfaces enforce the same policy.
    #[test]
    fn contract_fixtures() {
        let raw = include_str!("../../../fixtures/html-sanitize-cases.json");
        let parsed: Cases = serde_json::from_str(raw).expect("fixtures parse");
        for c in parsed.cases {
            let out = sanitize(&c.input);
            for needle in &c.must_contain {
                assert!(
                    out.contains(needle.as_str()),
                    "[{}] expected output to contain {needle:?}, got {out:?}",
                    c.name
                );
            }
            for needle in &c.must_not_contain {
                assert!(
                    !out.contains(needle.as_str()),
                    "[{}] expected output NOT to contain {needle:?}, got {out:?}",
                    c.name
                );
            }
        }
    }
}
